stdlib/internal/promql/join.go

```diff
@@ -0,0 +1,561 @@
+	if err := a.Do(func(cr flux.ColReader) error {
+		cr.Retain()
+		ar = append(ar, cr)
+		return nil
+	}); err != nil {
+		a.Done()
+		return nil, err
+	}
+	if err := b.Do(func(cr flux.ColReader) error {
+		cr.Retain()
+		br = append(br, cr)
+		return nil
+	}); err != nil {
+		a.Done()
+		b.Done()
+		return nil, err
+	}
+
+	// Release column readers after joining
+	defer a.Done()
+	defer b.Done()
+
+	timeColA := execute.ColIdx(execute.DefaultTimeColLabel, a.Cols())
+	if timeColA < 0 {
+		return nil, errors.New("no _time column found")
+	}
+
+	recordsA := records{
+		row:     -1,
+		col:     timeColA,
+		readers: ar,
+	}
+
+	timeColB := execute.ColIdx(execute.DefaultTimeColLabel, b.Cols())
+	if timeColB < 0 {
+		return nil, errors.New("no _time column found")
+	}
+
+	recordsB := records{
+		row:     -1,
+		col:     timeColB,
+		readers: br,
+	}
+
+	var (
+		timeA, timeB int64
+		firstRow     = true
+	)
+
+	rowA := make(map[string]values.Value, 8)
+	rowB := make(map[string]values.Value, 8)
+
+	idxA, readerA := recordsA.next()
+	idxB, readerB := recordsB.next()
+
+NEXT:
+	if readerA == nil || readerB == nil {
+		goto DONE
+	}
+
+	timeA = recordsA.timeValue()
+	timeB = recordsB.timeValue()
+
+	if timeA < timeB {
+		idxA, readerA = recordsA.next()
+		goto NEXT
+	}
+	if timeA > timeB {
+		idxB, readerB = recordsB.next()
+		goto NEXT
+	}
+	// _time values of both rows are equal
+	if timeA == timeB {
+
+		// Get the current row from each column reader
+		rowFromReader(readerA, idxA, rowA)
+		rowFromReader(readerB, idxB, rowB)
+
+		// Evaluate fn over both input rows
+		obj, err := c.fn.Eval(
+			c.ctx,
+			values.NewObjectWithValues(rowA),
+			values.NewObjectWithValues(rowB),
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		// Build schema if this is the first row being joined
+		if firstRow {
+			schema := make([]flux.ColMeta, 0, obj.Len())
+			obj.Range(func(name string, v values.Value) {
+				schema = append(schema, flux.ColMeta{
+					Label: name,
+					Type:  execute.ConvertFromKind(v.Type().Nature()),
+				})
+			})
+			sort.Slice(schema, func(i, j int) bool {
+				return schema[i].Label < schema[j].Label
+			})
+			for _, col := range schema {
+				if _, err := builder.AddCol(col); err != nil {
+					return nil, err
+				}
+			}
+			firstRow = false
+		}
+
+		// Check fn does not update the group key values.
+		// TODO(josh): This should be caught during planning.
+		// Remove this when the planner is made schema aware.
+		for _, col := range key.Cols() {
+			v, ok := obj.Get(col.Label)
+			if !ok {
+				return nil, errors.New("argument 'fn' may not modify group key")
+			}
+			if !v.Equal(key.LabelValue(col.Label)) {
+				return nil, errors.New("argument 'fn' may not modify group key")
+			}
+		}
+
+		// The record obtained from calling fn may be added to output
+		if err := appendRowToBuilder(builder, obj); err != nil {
+			return nil, err
+		}
+		idxB, readerB = recordsB.next()
+		goto NEXT
+	}
+DONE:
+	return builder.Table()
+}
```

Comment:
Kind of a corner case I'm curious about here: what if you had two tables A and B with the same group key and a bunch of rows in both sides with the same value in the `_time` column?  (I know that storage cannot produce such data, but it can exist in Flux.) . In that case I would expect a cross join for that subset of rows.  

E.g, if you had 10 rows in `A` with `_time=1000` and another 10 such rows in `B`, then there would be 100 corresponding rows in the join output.

It's unlikely to come up in practice very often now, but it would be good to have a test for it.  It will become more frequent if the join predicate were expressed as an arbitrary function passed in by the user.

Reply:
Thanks for catching this. I totally overlooked it.